[Feature] Generate flow events for every flow added or removed

viniarck opened the pull request:
Fixes #2  

### Description of the change

This PR added support to publish the rest of flows events to support reliably publishing events when flows are added, removed, still pending (not confirmed yet), and also handled switch connection error `kytos/core.openflow.connection.error` to bubble up the error to the caller. So, now `flow_manager` clients will be able to react accordingly and handle each event as needed based on their use case, not necessarily all clients will have to react and handle, but as they need, the events to completely handle every case are available, [handling the `kytos/flow_manager.flow.error` event is highly encouraged though at least, as `mef_eline` already does](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L664-665).

### Release notes

`version [5.3.0] - 2021-11.21`

#### Added

- Started listening to ``kytos/core.openflow.connection.error`` and propagating the error
- Added listen_to for ofpt_flow_removed
- Publish the event ``kytos/flow_manager.flow.removed`` on `OFPT_FLOW_REMOVED`
- Parametrized and added ``_id`` on stored flows to confirm flow operations
- Parametrized the flow on ``SwitchNotConnectedError`` exception
- Added support to update flow ``state`` and flows are confirmed by the consistency check

#### Changed

- Publish the event ``kytos/flow_manager.flow.added`` only when the flow is confirmed

viniarck commented on main.py:
```diff
@@ -87,6 +87,14 @@ def stored_flows_list(self, dpid):
         """Ordered list of all stored flows given a dpid."""
         return itertools.chain(*list(self.stored_flows[dpid].values()))
 
+    def stored_flows_by_state(self, dpid, state):
```
This kind of filtering here we can delegate and perform directly on a DB later as we move to a new back-end later on.

viniarck commented on main.py:
```diff
@@ -544,19 +618,39 @@ def _send_flow_mod(self, switch, flow_mod):
+            flow,
+            "error",
+            error_command=error_command,
+            error_exception=event.content.get("exception"),
```
@ajoaoff @italovalcy we might need to augment [`mef_eline`'s method `handle_flow_mod_error`](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L668-L674) - if we keep using the `'kytos/flow_manager.flow.error'` as we're using here to also bubble up the socket error of a flow mod install, in case the `error_exception` is set, it should be treated the same way as a `SwitchNotConnectedError` that results in an HTTP 424, because when this happens even though the consistency check will try to add or delete later on the client might want to try another path when applicable (we don't know for how long this might keep happening), and if clients don't handle this, then traffic could be lost because temporarily until the consistency check send again, network traffic might be using other flow entries that are part of the circuit. Let me know your preference on how to handle this. I haven't mapped an issue on `mef_eline` for this yet, let's evolve the discussion.

italovalcy replied:
@viniarck if I understood correctly, upon receiving a `kytos/flow_manager.flow.error` and  `error_command=='add'` (the original request was to add a flow), mef_eline will call `evc.remove_current_flows()` (https://github.com/kytos-ng/mef_eline/blob/master/main.py#L690) which is basically the same way it handles the `SwitchNotConnectedError` that results in an HTTP 424 (https://github.com/kytos-ng/mef_eline/blob/master/models/evc.py#L514). Then, after removing the currently flows, the EVC will be marked as `active == False` (and theoretically it is `enabled == True`), leaving to the consistency check reactivate it (https://github.com/kytos-ng/mef_eline/blob/master/main.py#L66-L70). Please let me know if you meant something different.

viniarck replied:
Right, @italovalcy, yeah, this current error handling approach on `mef_eline` should work as intended. 

Originally, I meant to bring the discussion to this point here when the `error_exception` is also set (in addition to having the `error_command=='add'`, maybe this would just be a temporary Socket failure, so it could be re-tried a few times or even try to use another path, **but as we currently have it's save to try to remove since `mef_eline` will use the newly iplemented `force` option as well**. 

Thinking about an enhancement though, If we need to try to find an alternative path we could try to enhance this part here. It comes down to try to find an alternative path, if any, or keep retrying, when this type of exception is raised. Or, arguably, to facilitate for clients, maybe on `flow_manager` we could also have some retries there for such cases when the `error_exception` happened, and then only raise this error after a few attempts, and then once clients got this event with the exception, it's safe to assume that retries aren't worth anymore, and either they try to find an alternative path or completely remove. Wdyt?

italovalcy replied:
Understood @viniarck, that is a good point! Thanks for bringing this enhancement possibility. IMHO, I think it is a good idea to have the flow_manager trying a few more times before reporting the error to the requestor. I support this idea. Tow points to take into consideration for this proposed approach: 1) how can we differentiate permanent and temporary errors? 2) what interval should we use between each try?

The reason for 1 is: most of the times I've seen errors like this is due to a switch reconnection (temporary error). Switch reconnection (during the installation of a flow) can be a good case where the retries approach fits better than just removing the current path and waiting for the next consistency check. In the order hand, permanent errors such as invalid action, invalid port, table full, etc. those are not worth to keep trying (if possible to differentiate those).

For 2, maybe we can reply an exponencial backoff approach starting with a few milliseconds. Example: wait 100ms and try again, wait 200ms and try again, finally wait 400ms and try again, give up and generate the `kytos/flow_manager.flow.error `.

viniarck replied:
Yeah, for the point number 1) `kytos/core.openflow.connection.error` events are potential temporary errors, and ultimately, as we currently have here, it'll send a `kytos/flow_manager.flow.error` and set the `error_exception`, that way subscribers can still different them from `OFPT_ERROR` which will also be sent as `kytos/flow_manager.flow.error` but without the `error_exception` key, and then subscribers can take actions accordingly as needed. 

But, maybe having a new event on flow_manager `kytos/flow_manager.connection.error` (this one being a higher level version of the `kytos/core.openflow.connection.error` with the retries, that way also not mixing contexts with the `kytos/flow_manager.flow.error` that should be exclusive for `OFPT_ERROR`s as it was meant to be originally) would probably make things easier to follow and understand and for subscribers to handle and differentiate, wdyt? 

For 2), yep, for sure, I think exponential short retries should fit this pretty well, and it wouldn't impact or add too much delay, allowing subscribers that need responsiveness to still have it. I appreciated your feedback and suggestions @italovalcy, I'll map this in an issue.

italovalcy replied:
Thanks for the comments, @viniarck. Yeah, having the retries approach configurable is the best way forward. Regarding the discussion about a new event on flow_manager, I rather prefer the first suggestion you gave: clients listening to `kytos/flow_manager.flow.error` and differentiating temp/permanent errors by looking at the `error_exception` (or even don't send a `kytos/flow_manager.flow.error` before running out of retries).

viniarck replied:
Sounds good @italovalcy, I've captured this requirement and refined info that we've discussed here https://github.com/kytos-ng/flow_manager/issues/56. thanks for you review and feedback.

ajoaoff commented on main.py:
```diff
@@ -87,6 +88,14 @@ def stored_flows_list(self, dpid):
+    def stored_flows_by_state(self, dpid, state):
+        """Get stored flows dict filter by a state."""
+        filtered_flows = {}
+        for entry in self.stored_flows_list(dpid):
+            if entry.get("state") and entry["state"] == state:
+                filtered_flows[entry["_id"]] = entry
+        return filtered_flows
```
```suggestion
        for entry in self.stored_flows_list(dpid):
            if entry.get("state") == state:
                filtered_flows[entry["_id"]] = entry
        return filtered_flows
```

viniarck replied:
I've just commit the suggestion, thank you.

ajoaoff commented on main.py:
```diff
@@ -152,6 +197,14 @@ def switch_flows_by_cookie(switch):
         return installed_flows
 
+    @staticmethod
+    def switch_flows_by_id(switch):
+        """Build switch.flows indexed by id."""
+        installed_flows = {}
+        for flow in switch.flows:
+            installed_flows[flow.id] = flow
+        return installed_flows
+
```
```suggestion
    def switch_flows_by_id(switch):
        """Build switch.flows indexed by id."""
        return {flow.id: flow for flow in switch.flows}

```

viniarck replied:
Good catch. I completely forgot to use a dict comprehension. I'll commit this suggestion @ajoaoff

italovalcy reviewed (approved):
Looks good to me. Thanks for the changes, @viniarck. If I understood correctly, the flow events for added and removed were already being sent by flow_manager. However, you PR enhances the reliability of the information provided in the event, so that a `flow_manager.flow.added` is only sent when the flows are confirmed to be installed (same for other events). Which means that it will not introduce extra overhead of events, but instead it will only send those events when they are actually trusted to be sent.

viniarck commented:
> Looks good to me. Thanks for the changes, @viniarck. If I understood correctly, the flow events for added and removed were already being sent by flow_manager. However, you PR enhances the reliability of the information provided in the event, so that a `flow_manager.flow.added` is only sent when the flows are confirmed to be installed (same for other events). Which means that it will not introduce extra overhead of events, but instead it will only send those events when they are actually trusted to be sent.

Exactly, @italovalcy. Also, clients could leverage this accordingly to have simpler consistency checks in the future, if they can trust and rely on `flow_manager` with its consistency check and all of confirmation/error related events, then they'll have all the cases to either confirm that what they sent was successfully added or not as these events happen and react accordingly. There are some edge cases that you mentioned in terms of restart and how some NApps write to the DB, but hopefully we can also enhance that part in a non distant future. Thanks for reviewing.
